### PR TITLE
4244-Kanban-Column-Menu

### LINKED ISSUE #4244: Update the Kanban column menu
# Current behavior

There is only one button "Edit from settings" that links to the object page in Settings

CleanShot 2024-02-29 at 16 04 49

Links to Object Settings:

CleanShot 2024-02-29 at 16 05 28

# Desired Behavior

There should be 3 buttons:
1- Edit (Should redirect to the field settings page instead of the object settings page)
2- Hide (Should hide the step)
3- Add Record (It should add a record in the column below)

CleanShot 2024-02-29 at 16 03 59

1- Should link to the corresponding Select Field settings instead of Object Settings
CleanShot 2024-02-29 at 16 06 42

# Figma

https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=19118-196797&mode=design&t=ZRf0c7Ey66nuhFnZ-11

### PR DESCRIPTION
**Pull request by Kanav-Arora**
Closes #4244

### CODE REVIEW

**Comment by Kanav-Arora** [comment]
Hi @charlesBochet 
I'm able to create new record but not in specific Stage because `RecordIndexBoardContainerEffect` is not in context provider of 

const { selectFieldMetadataItem } = useContext(RecordBoardContext);
  const { columnDefinition } = useContext(RecordBoardColumnContext);


therefore while creating new record I'm unable to provide selectFieldMetadataItem.name

File: packages/twenty-front/src/modules/object-record/utils/computeRecordBoardColumnDefinitionsFromObjectMetadata.ts [hunk 1]
```diff
@@ -34,11 +35,25 @@ export const computeRecordBoardColumnDefinitionsFromObjectMetadata = (
...
       {
         id: 'edit',
-        label: 'Edit from settings',
-        icon: IconPencil,
+        label: 'Edit',
+        icon: IconSettings,
         position: 0,
         callback: navigateToSelectSettings,
       },
+      {
```
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/utils/computeRecordBoardColumnDefinitionsFromObjectMetadata.ts:43]: Are we sure we're not duplicating something here ?

Comment (Kanav-Arora) [reply packages/twenty-front/src/modules/object-record/utils/computeRecordBoardColumnDefinitionsFromObjectMetadata.ts:43]: No no, I wanted to confirm if the hide functionality is already implemented or not. The hide function is to be replaced

**Comment by FelixMalfait** [comment]
Hey @Kanav-Arora not an easy one, thanks for taking it!

For the "hide" part we don't want to navigateToSelectSettings but instead we'd want to leverage views (viewFields) to save the visible columns like we do on the traditional list views.

As for your above question, could you do something like this? (not tested)
- Modify the callback in RecordBoardColumnAction to take an optional column id
- Modify computeRecordBoardColumnDefinitionsFromObjectMetadata to pass selectOption.id to the callback
- Modify createEmptyObjectWithStage to pass the stageId upon creation

**Comment by Kanav-Arora** [comment]
Hey @FelixMalfait 
I'll look into it

**Comment by FelixMalfait** [comment]
Linked discussion: https://discord.com/channels/1130383047699738754/1225049182839111690/1225049186387492964

**Comment by charlesBochet** [comment]
I have also commented on discord.
Pasting the discussion here for future record:

**Félix — Yesterday at 2:54 PM**
I think the right path would be to add a jsonb "options" or "hiddenOptions" column on the viewfield table. On that column we would store the fields that are explicitly hidden. Wdyt @charles? 
I know the behavior of storing the hidden options feels inconsistent with what we are doing at the field level where we have to explicite create the viewFields but it feels closer to the product POV where we want options to be available by default. OK for me if we keep the same behavior for consistency and have to update every options column when adding a new option value

**Kanav Arora — Yesterday at 4:35 PM**
I'm not much aware of the backend architecture but basically we are filtering out rows of data so can we maintain the id of stage which is hidden and filterout the rows with that id
sorry if I'm wrong

**Félix — Today at 10:21 AM**
You'd need to copy what is done for viewFields, but doing it at a second level which are options. The difference is that viewFields are stored in their own table while options would be a json within a viewField record to avoid table bloat
But overall it's very similar
I think we should copy the patterns exactly. Inverting the pattern as I initially suggested saying we should store the options we hide would be a bad pattern in the end I think

**charles — Today at 12:57 PM**
so we add a new field to the view object (RawJSON type): kanbanOptions which will contain: 
{ 
    optionVisibilities: [
       { optionId: string, visible: boolean }
    ]
} 
I'm not sure about the naming, right now columns of a kanban are selectOptions but we will authorize building kanban on top of other field type later
note that we have option.id which is more robust than option.label for SELECT but we still need to think about what's going to happen when we add / remove SELECT options in the settings
Alternatively, we can also introduce a table for that, but as options of SELECT are already store in JSON, we won't benefit from postgres constraint engine (to keep the data robust) so I don't really see a point justifying a new table. Both work, JSON is easier to implement

**Comment by FelixMalfait** [comment]
Bonus point: have a virtual column for items that don't have a select value than we can chose to display or not (cc @Thomas)
I found it strange that I couldn't find a card and it was because the select field was nullable and the card just didn't appear anywhere

**Comment by FelixMalfait** [comment]
Hey @Kanav-Arora, still working on this one? Or should we close the PR for now?

**Comment by charlesBochet** [comment]
@Kanav-Arora I'm closing this PR to keep the PR history clean but feel free to re-open it when you pick it again

**Comment by FelixMalfait** [comment]
Would be great if you do! Will be very cool to have this!

**Comment by charlesBochet** [comment]
Yes, happy to review it when it's ready!

**Pull request closed without merging**